pixeltable/exprs/function_call.py

```diff
@@ -300,8 +301,11 @@ def substitute(self, spec: dict[Expr, Expr]) -> Expr:
         """
         res = super().substitute(spec)
         assert res is self
-        self.return_type = self.fn.call_return_type(self.bound_args)
-        self.col_type = self.return_type
+        # See if we can update the return_type with something more specific.
+        crt = self.fn.call_return_type(self.bound_args)
+        if self.return_type.is_supertype_of(crt):
```

Comment:
Why is this needed? Shouldn't the result of call_return_type() be at least as specific as self.return_type?

Reply:
You're right, it should be, *as long as the codebase is stable*. But if (say) something changes in the implementation of call_return_type() then that's not guaranteed.

It took me some time to reason through all this, but I think I now understand it. I changed the condition here to `if self.is_valid`, which better captures what's actually going on, but it still isn't perfect. More details are in the new code comment. Getting the edge cases exactly right would involve some changes to EmbeddingIndex that require a migration, which probably isn't worth doing right now, since the PR as written fixes the proximate issue (graceful loading of metadata in cases where libraries needed by call_return_type() are missing).